_drafts/code-properties-and-language-features.md

```diff
@@ -0,0 +1,50 @@
+
+# Questioning Elixir and Redux
+
+First, I asked about the way data is separate from functionality in Elixir, and whether that caused problems. But I learned that when you define a struct inside a module, you also usually define functions that operate on that struct inside the same module. Because the name of the module and the name of the struct are the same, they go together naturally. So data and functionality aren't entirely separated.
+
+Then I asked about encapsulation: the data in a struct is available for everyone else in the app to access directly. But I learned that often you might have a convention of only accessing data in the struct via a function, and that that is a way to isolate knowledge of the data structure.
+
+Then I asked about [the `Plug.Conn` struct][conn] in the Phoenix framework, which is passed through much of the app. I asked about whether this made the entire application fragile, because all of it required intimate knowledge of the Connection. I learned that there was a [`private`][conn-private] field on the struct, and although you aren't prevented from accessing it directly, the convention is that you shouldn't because the implementation in it can change without warning. This reminded me of some Rubyists' point that, because you can get around the private keyword, it's effectively just a convention as well.
+
+After all this, I still had a concern with [Elixir's pattern matching][pattern-matching]. It's a lot like a switch statement, and [switch statements are a code smell][switch] in OO design. It makes it difficult to change your code because switch statements proliferate: every time you have to add a new case, all the switch statements need to be updated. Instead it's best to use polymorphism to handle variations, so that you can always add a new type of object and implement the new case's behavior on it. Each place you previously had a switch statement, you now send a message instead. But Elixir has a potential answer to this as well: [protocols][protocols] allow you to define a function signature, and then you can implement it for specific data types.
+
+I had a similar question about Redux. Having a single reducer you call with actions seemed to combine all the logic in one place. But it's conventional to [compose your reducer][reducer-composition] out of smaller reducers that produce the values for individual keys in your state. So each smaller reducer only has responsibility for one piece of state, and it can just return the data unchanged for any actions that don't pertain to it. Yes, there is just one big list of action names that are strings, but if you like you can introduce namespacing into your action names to group them by data type.
+
+At this point I was really confused. Were there *any* differences in capabilities between OO and FP? Had OO's grouping of data together with functionality really accomplished anything?
```

Comment:
Distributed OO between mutually untrusting nodes is a lot easier than distributed FP. Former was the Tech of the Day in production in 1990s with stuff like CORBA. Latter is still WIP AFAIK.